webpack/webpack-embedded-dependencies-plugin/src/EmbeddedDependenciesWebpackPlugin.ts

```diff
@@ -0,0 +1,361 @@
+/**
+ * @alpha
+ * Webpack plugin that generates a file with the list of embedded dependencies
+ * and their licenses.
+ */
+export default class EmbeddedDependenciesWebpackPlugin implements WebpackPluginInstance {
+  private readonly _outputFileName: string;
+  private readonly _generateLicenseFile: boolean;
+  private readonly _generateLicenseFileFunction: LicenseFileGeneratorFunction;
+  private readonly _generatedLicenseFilename: LicenseFileName;
+
+  public constructor(options?: IEmbeddedDependenciesWebpackPluginOptions) {
+    this._outputFileName = options?.outputFileName || 'embedded-dependencies.json';
+    this._generateLicenseFile = options?.generateLicenseFile || false;
+    this._generateLicenseFileFunction =
+      options?.generateLicenseFileFunction || this._defaultLicenseFileGenerator;
+    this._generatedLicenseFilename = options?.generatedLicenseFilename || 'THIRD-PARTY-NOTICES.html';
+  }
+
+  /**
+   * @alpha
+   * Webpack plugin apply method. This method is called by the webpack compiler to apply the plugin, however it not usually
+   * needed to be invoked manually by the developer in a webpack configuration. However, if you are calling this plugin (applying it from another plugin)
+   * you can call `plugin.apply(compiler)` to apply the plugin and invoke it.
+   * @param compiler - The webpack compiler instance.
+   */
+  public apply(compiler: Compiler): void {
+    // Tap into compilation so we can tap into compilation.hooks.processAssets
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation, { normalModuleFactory }) => {
+      const thirdPartyPackages: ThirdPartyPackageMap = new Map();
+
+      normalModuleFactory.hooks.module.tap(
+        PLUGIN_NAME,
+        (module, moduleCreateData: IWebpackModuleCreateData, resolveData) => {
+          const { resourceResolveData } = moduleCreateData;
+          const pkg: IPackageData | undefined = resourceResolveData?.descriptionFileData;
+          const filePath: string | undefined = resourceResolveData?.descriptionFileRoot;
+
+          if (pkg && filePath && filePath.includes('node_modules')) {
+            const key: PackageMapKey = makePackageMapKeyForPackage(pkg);
+            thirdPartyPackages.set(key, { dir: filePath, data: pkg });
+          }
+
+          return module;
+        }
+      );
+
+      compilation.hooks.processAssets.tapPromise(
+        { name: PLUGIN_NAME, stage: Compilation.PROCESS_ASSETS_STAGE_REPORT },
+        async (assets) => {
+          const packages: IPackageData[] = [];
+
+          Async.forEachAsync(Array.from(thirdPartyPackages), async ([, { dir, data }]) => {
+            const { name, version } = data;
+            let licenseSource: string | undefined;
+            const license: string | undefined = parseLicense(data);
+            const licensePath: string | undefined = await this._getLicenseFilePath(dir, compiler);
+            if (licensePath) {
+              licenseSource = await FileSystem.readFile(licensePath);
```

Comment:
```suggestion
              licenseSource = await FileSystem.readFileAsync(licensePath);
```

Reply:
Gracefully handle this file not existing?

Reply:
With this now wrapped in a `try/catch`, the error should propagate up to the catch. If we find a licensePath but fail to read it, this could be for permissions, or FS related problems and should just float this message up the consumer. 

Or do we also need a nested try/catch here?

Reply:
Resolved this conversation offline. This is good with the trycatch we have.